### PR TITLE
#12151: Generalize `max_pool2d` code into generic pool op

### PR DESCRIPTION
**Pull request by jerrysky3**
### Ticket
#12151

### Problem description
Per the discussion in #12151, we would like to share code between `max_pool2d` and the upcoming `avg_pool2d`. This PR moves and generalize the max_pool2d code (into a similar format as the reduction ops) so the common part can be reused for avg_pool2d later (#14268)

### What's changed
- Rename `ttnn/cpp/ttnn/operations/pool/maxpool` and files under to `ttnn/cpp/ttnn/operations/pool/generic`
- Change max_pool2d uop into generic pool uop (currently only supports max_pool2d)
- Expect unchanged behavior on max_pool2d

### Checklist
- [x] Post commit CI passes (https://github.com/tenstorrent/tt-metal/actions/runs/12078712479)
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/pool/generic/device/pool_op.hpp [hunk 1]
```diff
@@ -80,5 +84,5 @@ struct MaxPool2D {
 }  // namespace ttnn::operations
 
 namespace ttnn::prim {
-constexpr auto max_pool2d = ttnn::register_operation<"ttnn::prim::max_pool2d", ttnn::operations::pool::MaxPool2D>();
+constexpr auto pool2d = ttnn::register_operation<"ttnn::prim::pool2d", ttnn::operations::pool::Pool2D>();
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/pool/generic/device/pool_op.hpp:87]: I think `max_pool2d` as an operation still makes sense. It called to a pool2d device operation, passing MaxPool2D enum value. Unless @pavlejosipovic @mywoodstock prefer it this way, I think its best to correct this change to retain max_pool2d operation.

File: ttnn/cpp/ttnn/operations/pool/generic/generic_pools.cpp [hunk 2]
```diff
@@ -2,28 +2,47 @@
...
-                           std::array<uint32_t, 2> dilation,
-                           const std::optional<const MemoryConfig>& memory_config,
-                           const std::optional<const TensorMemoryLayout> applied_shard_scheme) {
+namespace {
+
+// Return a single bf16 init value for the pool type in u32 (packed in the least 16 bits)
+uint32_t get_bf16_pool_init_value(Pool2DType pool_type) {
+    float value;
+    switch (pool_type) {
+        case Pool2DType::MAX_POOL2D: value = -std::numeric_limits<float>::infinity(); break;
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/pool/generic/generic_pools.cpp:25]: I remember seeing that inf and none values are a bit special and might not directly map to std::numeric_limits.
OP devs know better. Just highlighting this. Will learn from the response to this comment.

Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/pool/generic/generic_pools.cpp:25]: See the old value that was used `uint32_t neg_inf_pad_val = 0xf7ff;`

File: ttnn/cpp/ttnn/operations/pool/generic/device/pool_op.hpp (hunk 1 above)
Comment (ayerofieiev-tt) [reply ttnn/cpp/ttnn/operations/pool/generic/device/pool_op.hpp:87]: Ah, nvm, I found the max_pool2d op declaration below.

**Review by ayerofieiev-tt — APPROVED** [review]
Leaving an approve from TT-NN CMake standpoint and overall change.
 
The change definitely needs a review/approve from the OP team and likely requires a correction around handling of `-inf/inf` values

File: ttnn/cpp/ttnn/operations/pool/generic/generic_pools.cpp (hunk 2 above)
Comment (jerrysky3) [reply ttnn/cpp/ttnn/operations/pool/generic/generic_pools.cpp:25]: Yes, we need to convert float32 to bfloat16 for using in kernel. This is done below when we return the value

```cpp
return bfloat16(value).to_packed();
```

The converted bfloat16 is also packed in u32 in order to pass in kernel compile arguments. In kernel we convert (truncate) u32 back to u16 to get the bfloat16. This behaivor is documented in the comment above of this function

**Review by pavlejosipovic — APPROVED** [review]
Nice change, thx for doing this step-by-step.

**Review by mywoodstock — APPROVED** [review]

**Pull request merged**